[ECAM] Implement FWC flight phases and update ECAM messages

beheh opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
Fixes #942
Fixes #1067
Fixes #1683
Fixes #1882

## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->
Per Asobo's default, the aircraft uses a unified, simple flight phase mechanism that is shared between all instruments (e.g. to differentiate takeoff vs. on ground behaviour). In some cases though, this is not detailled enough and the phases don't correlate with what the real aircraft does.

This pull request fully decouples the flight phases in the ECAM from the original flight phases, and implements the original FWC flight phase logic. This greatly enhances the accuracy of when certain ECAM messages appear, and in turn fixes some bugs like the T.O. memo appearing after landing. In future this should also simplify implementing new ECAM messages.

This PR doesn't fix nor implement every single ECAM message, but it's a stepping stone for more of that in future. It also enables a future PR to fully implement the ECAM message clear/recall/emerg cancel behaviour.

Detailed changelog for ECAM messages:
- OVERSPEED can no longer appear on the ground
- ENG FAIL can now appear on the ground in an appropriate flight phase
- BRAKES HOT no longer appears after landing until both engines are shut down for 5 minutes
- PARK BRK ON appears as an ECAM level 2 alert during flight (normal status only on ground)
- A/SKID N/WS OFF is inhibited during takeoff
- TCAS FAULT no longer appears during takeoff and landing
- IR IN ALIGN only appears before takeoff (amber during engine on not yet implemented) 
- LAND ASAP is inhibited during takeoff roll and landing roll
- T.O. INHIBIT and LDG INHIBIT show more appropriately now and have a 3 second delay
- SPEED BRK is inhibited after engine start and after landing
- AUTO BRK LO/MED/MAX is only shown when landing is imminent
- NW STRG DISC turns amber when an engine is running
- PRED W/S OFF turns amber during takeoff and landing
- COMPANY MSG is inhibited during takeoff and landing
- IGNITION added
- SIGNS...ON memo during takeoff memo now also requires NO SMOKING


## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->

https://youtu.be/qUNgmQcwGOw (slightly older version - T.O and LDG INHIBIT have a 3s delay now according to documentation)

## References
<!-- You should be making changes based on some kind reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->

![grafik](https://user-images.githubusercontent.com/929769/99184532-c701ad00-2743-11eb-8e50-b6188521f7b2.png)

![image](https://user-images.githubusercontent.com/929769/99393359-82aa1480-28dd-11eb-8053-c57983064a3d.png)

![image](https://user-images.githubusercontent.com/929769/99393370-89388c00-28dd-11eb-9d6f-2b404da77ba3.png)


<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

## Additional context
<!-- Add any other context about the pull request here. -->

### QA

For this one, watch the ECAM apart from the usual testing feel free to compare appearing messages various real sources.
While some messages are still missing as before, most of the ones that appear should only appear in an appropriate flight pase now. Various ECAM messages have also been subtly changed according to the docs, adjusting when they appear and in some cases how they appear (e.g. NW STRG DISC in amber while an engine is running).

Specifically If you encounter unexpected ECAM messages or an unexpected T.O/landing memos please take a note of the specific message, and describe how to reproduce it.

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

Benjozork commented:
Ate there any other instruments where we can reuse the same phase definitions ?

beheh commented:
Hm, I don't think so. AFAIK the FMA is mostly driven by the FMC (which are mostly correct right now, but I have a little more cleanup planned there after #1886), and I can't think of anything else obvious that uses these.

Benjozork reviewed (approved):
LGTM

beheh commented:
Keeping as draft for a bit while I'm investigating why the lower ecam runs the upper ecam code. Should not have a functional impact, it's just weird.

beheh commented:
Waiting on #1965 to land.

beheh commented:
Okay, #1965 is written, but I don't think that blocks the work here. Both PRs can land independently.

beheh commented:
I feell ike I should move the FWC to A32NX_Core instead of the ECAM, so back to draft for a bit.

beheh commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.html:
```diff
@@ -20,6 +20,7 @@
 
 <script type="text/html" import-script="/JS/dataStorage.js"></script>
 <script type="text/html" import-script="/Pages/A32NX_Utils/NXDataStore.js"></script>
+<script type="text/html" import-script="/Pages/A32NX_Core/A32NX_FWC.js"></script>
```
I'm not super happy about importing thise here. An alternative would be to move the logic nodes to A32NX_Utils, but I wasn't sure about the convention and how it would be used.

nistei replied:
The logic nodes seem to be incredibly versatile. I would suggest moving them to the Utils.

beheh replied:
Sounds good. Any suggestion for a file name? I see the previous ones are in some kind of NX namespace - or is that somethnig else? How about `NXLogic.js`?

nistei replied:
Sounds good to me.

beheh replied:
Done.

InhumanReactions commented:
Category: QA Tester
Name: InhumanReactions #7129
Date of testing: Current date 17/11/2020
Version of the sim: 1.10.11.0
PR Tested: #1943
Object of test: [ECAM] Implement FWC flight phases and update ECAM messages

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. Spawned in C&D at EGLL and set up the aircraft as standard.
2. Pushed back and the PRED W/S switched to being orange. Switched back to green after starting Engine #2 (not sure if this is supposed to happen).
3. Departed from EGLL. T/O Memo disappeared as expected when TOGA was activated. T.O Inhibit also activated with TOGA so all is working well. 
4. Flew the rest of the route up to EGGP
5. Landed normally, the Landing Memo came up at the correct time, and the Landing Inhibit memo also came up correctly.
6. Afterwards I loaded onto the runway, however the IRS wasn't aligned (for some unknown reason). The ECAM memos were working as listed. I attempted again, and the IRS worked this time and were aligned. Not sure if that is a problem or a one off. However both times the Signs switch wouldn't register movement of the switch and after being switched off, the ECAM displayed it as being on.

Flight notes (If required): Successful flight.

Issues (If Any): The Signs memo wouldn't register physical movement, even after flicking the switch multiple times. It also wouldn't register on landing, again I tried flicking the switch on and off. Even when starting on the runway, the changes still wouldn't register on the Upper ECAM. One last thing, the Lower ECAM display didn't switch from ENGINE to CRUISE throughout the flight. The phases in the MCDU were correct, but the ECAM stayed on ENGINE until I landed it switched to F/CTR and then subsequently APU when I started it up. Finally, it switched to the DOOR/OXY page and stayed there as usual. To test this load in C&D and go through a full flight. You may run into this issue too. After testing a go-around and selected heading, I can confirm the CRUISE page does work then. To reproduce it during a full flight, I would recommend starting C&D, as that is what I did and it didn't work.
Overall Rating: To be improved
Conclusions of the testing: This PR is really good aside from the issues mentioned above. Just the Lower ECAM not switching to CRUISE and the SIGNS memo not registering physical movement of the switch. Also the PRED W/S may not be fully working, but that may be how it is supposed to be, I am not sure. But yeah, good PR, just a few minor issues experienced. Good work! (Also sorry if this is too long, just a new tester and learning the most efficient ways to writing these up!)

beheh commented:
Hey, thanks for taking my PR as your first QA session! Your test the feedback sound nice and detailed.

> Pushed back and the PRED W/S switched to being orange. Switched back to green after starting Engine #2 (not sure if this is supposed to happen).

That sounds like a bug. It should only turn amber once T.O. CONFIG is pressed (and then stay amber until takeoff). I'll look at it.

> the Lower ECAM display didn't switch from ENGINE to CRUISE 

I'm investigating this.

> However both times the Signs switch wouldn't register movement of the switch and after being switched off, the ECAM displayed it as being on.

As per the ESLD I've updated this memo item to require both the seatbelts to be on and the no smoking sign to be either on or auto. Is it possible you left the no smoking sign off (the default)?

beheh commented:
>Pushed back and the PRED W/S switched to being orange. Switched back to green after starting Engine #2 (not sure if this is supposed to happen).

While I'm unable to really reproduce this, the logic is a bit tricky and not quite correct the way it's right now. It should turn amber with at least one engine started and the T.O. Config button pressed, but that state and it's resets aren't fully mapped yet. For now I'm changing it so it's green until you're TOGA or in the air (at which point it's always amber, as per the docs).

> the Lower ECAM display didn't switch from ENGINE to CRUISE

I also can't reproduce this, and have a theory for this: You may have accidentally left your enginer starter switch near the pedestal on "IGN START". That forces the engine page. While I can't fix that, I've added the IGNITION message to the upper ECAM now as per the docs to remind you that you've left it on :-)

I believe that should address all the feedback from the test.

beheh commented:
I've gotten hold a more recent ESLD, so I've slightly tweaked ECAM message logic according to it. Would be good to get at least one more QA tester doinga full flight with the latest version.

beheh commented:
I've rebased. On more QA Test would be good. I've also added a detailed changelog for the ECAM messages to the PR here itself (but too verbose for the actual CHANGELOG).

CaptainR3X commented:
Category: QA Tester
Name: GustavRex#4347
Date of testing: Current date 19/11/2020
Version of the sim: 1.10.11.0
PR Tested: #1943
Object of test: Implement FWC flight phases and update ECAM messages

Testing technique: Flight; EKYT-EKCH
Steps you've made to test the Pull Request:
1. C/D at EKYT, normal programming procedures, p/s ECAM performed normally
2. Taxi and lineup, tested that NW STRG DISC performed as it should, departure 
3. Cruise, checking ECAM functionality
4. Arrival programming, following same procedure as in the beginning
5. Appproach and short final (NRG STRG DISC, Com. msg and ECAM performance testing during landing), GA, approach & landing (same procedures as first attempt) taxi and shut-down

Flight notes (If required): Online, VatSim - multiple DIR and reprogramming, Go-Around 

Issues (If Any): None related to this PR but bug found in FCU (will post on discord)
Overall Rating: Good
Conclusions of the testing: Testing went fine. The ECAM performed as it should, even when trying to provoke failures. I tested many of the points from the changelog and was unable to find abnormal performances even after go-around. The ECAM was fully functional.

wpine215 commented:
This needs its own separate issue, but from looking at the ESLD, "IGNITION" shouldn't appear for CFM engines during normal operations, even when the engine selector switch is in ignition. Also, not sure about "IRS ALIGNED" message - perhaps a real pilot can confirm.

beheh commented:
I'll get real pilot confirmation on the IRS ALIGNED - I went by this: 

![image](https://user-images.githubusercontent.com/929769/99779946-1b2ad980-2b16-11eb-93e7-1e4513079932.png)

...although looking at a reference more closely, it should a) disappear when either engine is running and b) according to a detailed reference it should say "IR ALIGNED", which makes more sense consistency-wise. I'll get confirmation and fix this in an upcoming ECAM PR.

Regarding CFM engines, I have this IGNITION memo reference here:

![image](https://user-images.githubusercontent.com/929769/99780191-6a710a00-2b16-11eb-8125-50d38044bbc3.png)

...but that refers to continuous ignition (and that's for all of PW/IAE/CFM). At the end of the day this would probably be implemented at that granularity in the FADEC, but I'll also check in with pilots if it appears during normal operations, then I'll also clean this up in an upcoming ECAM PR.

beheh commented:
@wpine215 I've filed #2029 to track my work on the issues you raised.
